# Preprocessing

## Loading our data from S3

In [0]:
import pyspark
sc = spark.sparkContext
from pyspark.sql import functions as F
from pyspark.sql.functions import isnull, when, count, col

In [0]:
filepath = "s3://full-stack-bigdata-datasets/Big_Data/youtube_playlog.csv"
ACCESS_KEY_ID = "AKIA3V3GLDX54DRFGYTB" # cle du compte student
SECRET_ACCESS_KEY = "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" # secret key du compte student

hadoop_conf = spark._jsc.hadoopConfiguration()
hadoop_conf.set("fs.s3a.access.key", ACCESS_KEY_ID)
hadoop_conf.set("fs.s3a.secret.key", SECRET_ACCESS_KEY)
hadoop_conf.set("fs.s3a.impl","org.apache.hadoop.fs.s3a.S3AFileSystem") 

In [0]:
# TODO: load the the into a PySpark DataFrame: `playlog`
### BEGIN STRIP ###
playlog = (spark.read.format('csv')\
             .option('header', 'true')\
             .option('inferSchema', 'true')\
             .load(filepath))
### END STRIP ###
playlog.show(5)

+----------+----+-----------+
 timestamp|user| song|
+----------+----+-----------+
1392387533| 0|t1l8Z6gLPzo|
1392387538| 1|t1l8Z6gLPzo|
1392387556| 2|t1l8Z6gLPzo|
1392387561| 3|we5gzZq5Avg|
1392387566| 4|we5gzZq5Avg|
+----------+----+-----------+
only showing top 5 rows

## First analysis

In [0]:
# TODO: print out our DataFrame's schema
### BEGIN STRIP ###
playlog.printSchema()
### END STRIP ###

root
-- timestamp: integer (nullable = true)
-- user: integer (nullable = true)
-- song: string (nullable = true)

In [0]:
# TODO: use `.describe(...)` on your DataFrame
### BEGIN STRIP ###
playlog.describe().toPandas()
### END STRIP ###

summary 
 timestamp 
 user 
 song 
 
 
 
 
 0 
 count 
 25739537 
 25739537 
 25739537 
 
 
 1 
 mean 
 1.4427006561045842E9 
 12697.352275450798 
 2.532571778181818E8 
 
 
 2 
 stddev 
 3.443284872371195E7 
 13094.065905828476 
 8.334645614940468E8 
 
 
 3 
 min 
 -139955897 
 0 
 ---AtpxbkaE 
 
 
 4 
 max 
 1554321113 
 45903 
 zzzcFgRMY6c

### Missing values check

In [0]:
# TODO: count the missing values for each column
#       put the result in a pandas DataFrame and print it out
### BEGIN STRIP ###

playlog \
    .select([count(when(isnull(c), c)).alias(c) for c in playlog.columns]) \
    .toPandas()

### END STRIP ###

timestamp 
 user 
 song 
 
 
 
 
 0 
 0 
 0 
 0

### Duplicates check

In [0]:
playlog_count = playlog.count()

In [0]:
# TODO: check if playlog without duplicates has the same number of rows as the original
### BEGIN STRIP ###
val_distinct = playlog.distinct().count()
# OR val_distinct = playlog.select(countDistinct('timestamp', 'user', 'song'))
val_distinct == playlog.count()
### END STRIP ###

Out[6]: False

In [0]:
val_distinct

Out[84]: 25615886

Seems like we have duplicates, let's count how many.

In [0]:
# TODO: figure out a way to count the number of duplicates
### BEGIN STRIP ###
playlog_duplicate = playlog_count - val_distinct
playlog_duplicate
### END STRIP ###

Out[12]: 123651

In [0]:
# BONUS: compute descriptive statistics for duplicates 
#        How many duplicates on average (for duplicated rows)?
#        And the standard deviation?
#        What' the min? The max? 
#
# WARNING: this might be difficult
### BEGIN STRIP ###
playlog.groupBy(playlog.columns) \
    .count() \
    .where(F.col('count') > 1) \
    .describe('count') \
    .withColumnRenamed('count', 'dupli_desc_stats') \
    .toPandas()
### END STRIP ###

summary 
 dupli_desc_stats 
 
 
 
 
 0 
 count 
 80352 
 
 
 1 
 mean 
 2.538866487455197 
 
 
 2 
 stddev 
 1.653928906381983 
 
 
 3 
 min 
 2 
 
 
 4 
 max 
 66

### Other checks

In [0]:
# TODO: order the dataframe by ascending `timestamp` and show the first 5 rows
### BEGIN STRIP ###
playlog \
    .orderBy(playlog.timestamp.asc()) \
    .show(5)
### END STRIP ###

+----------+----+-----------+
 timestamp|user| song|
+----------+----+-----------+
-139955897| 4|nRa-eGzpT6o|
1392387533| 0|t1l8Z6gLPzo|
1392387537| 70|VJ6ofd0pB_c|
1392387537| 22|Q24VZL8wpOM|
1392387538| 1|t1l8Z6gLPzo|
+----------+----+-----------+
only showing top 5 rows

Do you see anything suspicious?

---

The first timestamp is negative, and it seems like it's the only one.  
We will make sure there aren't other like this.

In [0]:
# TODO: count the number of rows with a negative timestamp
### BEGIN STRIP ###
playlog \
    .select('timestamp') \
    .where(F.col('timestamp') < 0)\
    .count() 
    
### END STRIP ###

Out[71]: 1

As expected, only one such negative timestamp. Since we have only one we can actually `.collect(...)` it.

In [0]:
# TODO: collect the problematic rows
### BEGIN STRIP ###
playlog \
    .where(F.col('timestamp') < 0) \
    .collect()
### END STRIP ###

Out[74]: [Row(timestamp=-139955897, user=4, song='nRa-eGzpT6o')]

There's only one problematic value among more than 25M.  This negative timestamp is an error, as such the real value is missing. We could try to reconstruct the real value but that would be a really tedious task, since it's one value over 25M, we will simply remove it.

## Removing the row with a negative timestamp

We will use our new knowledge about the data to perform some preprocessing.  

Our pipeline will have 2 steps:
1. Remove duplicates (123651 rows)
2. Remove row with negative timestamps (1 row)

We will call our new DataFrame `playlog_processed` and save it to S3 in parquet format.

In [0]:
# TODO: filter out:
#       - duplicated values
#       - rows with negative timestamp
#       and save the result to a new DataFrame: `playlog_processed`
#       Finally, print out the number of rows in this DataFrame
### BEGIN STRIP ###
playlog_processed = playlog.distinct() \
                                 .filter(F.col('timestamp') > -1) 
playlog_processed.count()

### END STRIP ###

Out[10]: 25615885

In [0]:
# TODO:
ACCESS_KEY_ID = "****"# enter your S3 access key
SECRET_ACCESS_KEY = "****" # enter your s3 scecret key

hadoop_conf = spark._jsc.hadoopConfiguration()
hadoop_conf.set("fs.s3a.access.key", ACCESS_KEY_ID)
hadoop_conf.set("fs.s3a.secret.key", SECRET_ACCESS_KEY)
hadoop_conf.set("fs.s3a.impl","org.apache.hadoop.fs.s3a.S3AFileSystem") 

### END STRIP ###
output_path = 's3://full-stack-bigdata-datasets-1986/Big_Data/playlog_processed.parquet'

# TODO: save the processed DataFrame to S3 using the parquet format
### BEGIN STRIP ###
playlog_processed.write \
  .parquet(output_path, mode='overwrite')
### END STRIP ###

In [0]:
playlog_processed.write \
  .parquet(output_path, mode='overwrite')